experimental/harfbuzz/src/lib.rs

```diff
@@ -0,0 +1,324 @@
+    let general_category_map =
+        Box::new(icu_properties::maps::load_general_category(data_provider)?);
+    // TODO(#2833): mirroring
+    // TODO(#2832): script
+    let canonical_composition = Box::new(CanonicalComposition::try_new_unstable(data_provider)?);
+    let canonical_decomposition =
+        Box::new(CanonicalDecomposition::try_new_unstable(data_provider)?);
+    unsafe {
+        let empty = hb_unicode_funcs_get_empty();
+        // The HarfBuzz refcounting convention is that "create"
+        // sets refcount to one, not zero.
+        // https://harfbuzz.github.io/object-model-lifecycle.html
+        let ufuncs = hb_unicode_funcs_create(empty);
+        if ufuncs == empty {
+            return Err(HarfBuzzError::Alloc);
+        }
+        // Below this point, the only return is upon success. Now we
+        // can turn the `Box`es into raw pointers. Let's do that
+        // as an intermediate step with explicit types to be convinced
+        // that the types before casting to `*mut c_void` are the same
+        // ones that we'll cast back to in the callbacks.
+        let canonical_combining_class_map_ptr: *mut CanonicalCombiningClassMap =
+            Box::into_raw(canonical_combining_class_map);
+        let general_category_map_ptr: *mut CodePointMapData<GeneralCategory> =
+            Box::into_raw(general_category_map);
+        // TODO(#2833): mirroring
+        // TODO(#2832): script
+        let canonical_composition_ptr: *mut CanonicalComposition =
+            Box::into_raw(canonical_composition);
+        let canonical_decomposition_ptr: *mut CanonicalDecomposition =
+            Box::into_raw(canonical_decomposition);
+
+        hb_unicode_funcs_set_combining_class_func(
+            ufuncs,
+            Some(icu4x_hb_unicode_combining_class),
+            canonical_combining_class_map_ptr as *mut c_void,
+            Some(icu4x_hb_unicode_combining_class_destroy),
+        );
+        hb_unicode_funcs_set_general_category_func(
+            ufuncs,
+            Some(icu4x_hb_unicode_general_category),
+            general_category_map_ptr as *mut c_void,
+            Some(icu4x_hb_unicode_general_category_destroy),
+        );
+
+        // TODO(#2833):
+        // hb_unicode_funcs_set_mirroring_func();
+        // TODO(#2832):
+        // hb_unicode_funcs_set_script_func();
+
+        hb_unicode_funcs_set_compose_func(
+            ufuncs,
+            Some(icu4x_hb_unicode_compose),
+            canonical_composition_ptr as *mut c_void,
+            Some(icu4x_hb_unicode_compose_destroy),
+        );
+        hb_unicode_funcs_set_decompose_func(
+            ufuncs,
+            Some(icu4x_hb_unicode_decompose),
+            canonical_decomposition_ptr as *mut c_void,
+            Some(icu4x_hb_unicode_decompose_destroy),
+        );
+
+        // Compatibility decomposition and East Asian Width lookups
+        // are deprecated, and there's no need to set up the callbacks
+        // for those.
```

Comment:
Note: I don't even see setters for these in the Harfbuzz source.

Reply:
See `hb-deprecated.h`